pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,331 @@
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		options:           options,
+	}
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+// GetProxyServiceInstances returns service instances co-located with a given proxy
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
+		notReadyIP, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		notReadyPort, err := strconv.Atoi(port)
+		if proxyIP != notReadyIP || err != nil {
+			continue
+		}
+		se, ok := conf.Spec.(*networking.ServiceEntry)
+		if !ok {
+			return nil, errors.New("getProxyServiceInstances: wrong type")
+		}
+
+		for _, h := range se.Hosts {
+			for _, svcPort := range se.Ports {
+				// add not ready endpoint
+				out = append(out, d.notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
+				// add other endpoints
+				out = append(out, d.serviceInstances(se, svcPort, conf, h)...)
+			}
+		}
+	}
+	return out, nil
+}
+
+// GetService implements a service catalog operation by hostname specified.
+func (d *MCPDiscovery) GetService(hostname host.Name) (*model.Service, error) {
+	var uid string
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for _, conf := range d.notReadyEndpoints {
+		h := kube.ServiceHostname(conf.Name, conf.Namespace, d.options.DomainSuffix)
+		if hostname == h {
+			se, ok := conf.Spec.(*networking.ServiceEntry)
+			if !ok {
+				return nil, errors.New("getSerivce: wrong type")
+			}
+			if d.options.mixerEnabled() {
+				uid = fmt.Sprintf("kubernetes://%s.%s", conf.Name, conf.Namespace)
+			}
+
+			return &model.Service{
+				Hostname:   h,
+				Ports:      convertServicePorts(se.Ports),
+				Resolution: model.Resolution(int(se.Resolution)),
+				Attributes: model.ServiceAttributes{
+					Name:      conf.Name,
+					Namespace: conf.Namespace,
+					UID:       uid,
+				},
+			}, nil
+		}
+	}
+	return nil, errors.New("getSerivce: service not found")
+}
+
+// InstancesByPort implements a service catalog operation
+func (d *MCPDiscovery) InstancesByPort(svc *model.Service, servicePort int, labels labels.Collection) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
```

Comment:
This loop is not efficient, I suggest change `notReadyEndpoints ` to map[addr][port]conf